client/cmdhfmf.c

```diff
@@ -1530,6 +1557,423 @@ static int CmdHF14AMfNestedHard(const char *Cmd) {
+    }
+
+    t1 = msclock();
+
+    // If no key is supplied by the user, brute force with the dictionary
+    if (know_target_key == false) {
+        for (current_sector_i=0; current_sector_i < sectorsCnt; current_sector_i++) {
+            for (current_key_type_i=0; current_key_type_i < 2; current_key_type_i++) {
+                for (default_keys_i=0; default_keys_i < keycnt; default_keys_i++) {
+                    // Iterate over the keys
+                    for (i=0; i<6; i++) {
+                        tmpKey[i] = keyBlock[i + (6*default_keys_i)];
+                    }
+
+                    if (mfCheckKeys(current_sector_i*4, current_key_type_i, true, 1, tmpKey, &key64) == PM3_SUCCESS) {
+                        PrintAndLogEx(SUCCESS, "Jackpot, we found a key! Now let the fun begin!");
+                        PrintAndLogEx(SUCCESS, "[Dictio. KEYS] Valid KEY FOUND: sector:%3d key type:%c key: " _YELLOW_("0x%02x%02x%02x%02x%02x%02x"), 
+                            current_sector_i,
+                            current_key_type_i ? 'B' : 'A',
+                            tmpKey[0], tmpKey[1], tmpKey[2], tmpKey[3], tmpKey[4], tmpKey[5]);
+                        // Store the new key
+                        for (i=0; i<6; i++) {
+                            key[i] = tmpKey[i];
+                        }
+                        know_target_key = true;
+                        blockNo = current_sector_i;
+                        keyType = current_key_type_i;
+                        foundKeysDictionary++;
+
+                        // Exit the loop
+                        current_sector_i = sectorsCnt;
+                        current_key_type_i = 2;
+                        default_keys_i = keycnt;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
+    // Set the user defined / bruteforced key
+    if (know_target_key) {
+        e_sector[blockNo].Key[keyType] = bytes_to_num(key, 6);
+        arr[blockNo + (keyType * sectorsCnt)] = 1;
+    } else {
+        PrintAndLogEx(FAILED, "No usable key was found!");
+        free(e_sector);
+        return 1;
+    }
+
+    // Iterate over each sector and key(A/B)
+    for (current_sector_i=0; current_sector_i < sectorsCnt; current_sector_i++) {
+        for (current_key_type_i=0; current_key_type_i < 2; current_key_type_i++) {
+
+            foundkey = e_sector[current_sector_i].Key[current_key_type_i];
+
+            // Try the found keys
+            if (foundkey == 0) {
+                for (found_keys_i=0; found_keys_i < sectorsCnt; found_keys_i++) {
+                    // Iterate over the keys
+                    if (arr[found_keys_i + (current_key_type_i * sectorsCnt)] == 1) {
+                        num_to_bytes(e_sector[found_keys_i].Key[current_key_type_i], 6, tmpKey);
+                        if (mfCheckKeys(current_sector_i*4, current_key_type_i, true, 1, tmpKey, &key64) == PM3_SUCCESS) {
```

Comment:
same as above

Reply:
This is more reliable and fast enougth. The fast key check looks a bit buggy when it's used with only one key (it returned mutiple times the wrong keys).